HULKs/hulk
# Visual referee

## Introduced Changes

Implement Visual Referee for new SPL rules. It allows the transition from Initial to Ready by detecting if the head referee stands on the opposite side of the technical area and behind the T-Intersection and raises their arms above their heads.  

Fixes #549 
Based on #769 
Based on #809 
## ToDo / Known Issues

- [x] Deactivating the pose detection in any state other than Initial
- [x] Correctly set the expected referee position when on away side
- [x] Introduces heavy stuttering to control when used in combination with  #769 
- #809 Fixes the stuttering
- [x] #809 unfortunately also introduces a head wobble while trying to `look_at` some `Ground` point.

## Ideas for Next Iterations (Not This PR)

- [ ] If NAOs are penalized, the minimum amount of detection for the decision should be dynamically changed #816 
- [ ] Implement pose detection filter to make sure at least some amount of successive detections occur before commiting to the detection #818 
- [ ] Reduce the amount of network messages being sent #817 
- [ ] Use a cropped image which is not downsampled and use that for the pose inference 

## How to Test

1. Make sure the camera is calibrated, otherwise the position of the detected poses may not be accurate enough.
2. Deploy to NAOs with accurate player number according to their Initial position.
3. Start Game Controller and bring game into Initial state.
4. Go to referee position and hold arms over head. 
5. If at least two NAOs recognized the pose, all NAOs should go to Ready state and start moving. 
6. For validation of detected referee pose position and type, use twix Image and Map overlays. 
7. Repeat the procedure for the Initial to Ready transition in the second half.
